API/hermes_abi/hermes_abi.h

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#ifndef HERMES_ABI_HERMES_ABI_H
+#define HERMES_ABI_HERMES_ABI_H
+
+#include <assert.h>
+#include <stdalign.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#ifndef HERMES_EXPORT
+#ifdef _MSC_VER
+#define HERMES_EXPORT __declspec(dllexport)
+#else // _MSC_VER
+#define HERMES_EXPORT __attribute__((visibility("default")))
+#endif // _MSC_VER
+#endif // !defined(HERMES_EXPORT)
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/// Define the structure for references to pointer types in JS (e.g. string,
+/// object, BigInt).
+/// TODO: Replace jsi::PointerValue itself with this C implementation to
+/// eliminate pointer management overhead in the JSI wrapper.
+struct HermesABIManagedPointer;
+struct HermesABIManagedPointerVTable {
+  /// Pointer to the function that should be invoked when this reference is
+  /// released.
+  void (*invalidate)(HermesABIManagedPointer *);
+};
+struct HermesABIManagedPointer {
+  const HermesABIManagedPointerVTable *vtable;
+};
+
+enum HermesABIErrorCode {
+  HermesABIErrorCodeJSINativeException,
+  HermesABIErrorCodeJSError,
+};
+
+static_assert(
+    alignof(HermesABIManagedPointer) % 4 == 0,
+    "HermesABIManagedPointer must be at least aligned to pointer size.");
+
+/// Define simple wrappers for the different pointer types.
+#define HERMES_ABI_POINTER_TYPES(V) \
```

Comment:
While the macros are easier to write, they are very difficult to debug and to search for symbols.
It would be better to write all these definitions explicitly.
The API is usually written once and read many times.
Being explicit here significantly increases readability of the code.